File: I1_MVPAndRetrofit_Kotlin/app/src/main/java/com/example/tsvetelinastoyanova/weatherapp/data/source/local/AppDatabase.java
package com.example.tsvetelinastoyanova.weatherapp.data.source.local;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.content.Context;

import com.example.tsvetelinastoyanova.weatherapp.Constants;
import com.example.tsvetelinastoyanova.weatherapp.data.CityEntity;

@Database(entities = {CityEntity.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    public abstract CityDao cityDao();

    private static AppDatabase INSTANCE;

    private static final Object sLock = new Object();

    public static AppDatabase getInstance(Context context) {
        synchronized (sLock) {
            if (INSTANCE == null) {
                INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                        AppDatabase.class, Constants.Companion.getDATABASE_NAME())
                        .build();
            }
            return INSTANCE;
        }
    }
}

File: I1_MVPAndRetrofit_Kotlin/app/src/main/java/com/example/tsvetelinastoyanova/weatherapp/data/source/local/AppDatabase.kt
package com.example.tsvetelinastoyanova.weatherapp.data.source.local

import android.arch.persistence.room.Database
import android.arch.persistence.room.Room
import android.arch.persistence.room.RoomDatabase
import android.content.Context
import com.example.tsvetelinastoyanova.weatherapp.Constants
import com.example.tsvetelinastoyanova.weatherapp.data.CityEntity

@Database(entities = [CityEntity::class], version = 1)
abstract class AppDatabase : RoomDatabase() {

    abstract fun cityDao(): CityDao

    companion object {

        @Volatile
        private var INSTANCE: AppDatabase? = null

        fun getInstance(context: Context): AppDatabase =
                INSTANCE ?: synchronized(this) {
                    INSTANCE ?: buildDatabase(context)
                            .also { INSTANCE = it }
                }

        private fun buildDatabase(context: Context) =
                Room.databaseBuilder(context.applicationContext,
                        AppDatabase::class.java, Constants.DATABASE_NAME)
                        .build()
    }
}